### PR TITLE
feat: (ui) input and entity types mapping indicator

### PR DESCRIPTION
**Pull request by lordmallam**
Indicates which properties have been mapped on `input` and `entity types`

#### Related JIRA ticket
[AET-230](https://jira.ehealthafrica.org/browse/AET-230)

#### To be styled

### CODE REVIEW

File: aether-ui/aether/ui/assets/apps/pipeline/sections/Mapping.jsx [hunk 1]
```diff
@@ -12,12 +12,33 @@ class Mapping extends Component {
     this.state = {
       mappingRules: this.parseProps(props)
     }
+    this.setMappingStyles()
   }
 
   componentWillReceiveProps (nextProps) {
+    if (nextProps.selectedPipeline.mapping !== this.props.selectedPipeline.mapping) {
```
Comment (obdulia-losantos) [inline aether-ui/aether/ui/assets/apps/pipeline/sections/Mapping.jsx]: I found out that this line is most of the time true when it shouldn't (pointers and immutability) so it's better to use the `deepEqual` method that checks content not pointers.

`if (!deeptEqual(nextProps.selectedPipeline.mapping, this.props.selectedPipeline.mapping)) {`

**Comment by obdulia-losantos** [comment]
In the input side, shouldn't the rest of fields in the mapping rules be bold as the entity type ones are?

image

**Comment by sabeen** [comment]
you mean that _forename_ etc. should also have a background, like the __id_?
Yes, it should.

File: aether-ui/aether/ui/assets/apps/pipeline/sections/Mapping.jsx [hunk 2]
```diff
@@ -12,12 +12,33 @@ class Mapping extends Component {
...
+      this.clearMappingStyles(this.props.selectedPipeline.mapping)
+    }
     this.setState({
       mappingRules: this.parseProps(nextProps)
     })
+    this.setMappingStyles()
+  }
+
+  setMappingStyles () {
```
Comment (obdulia-losantos) [inline aether-ui/aether/ui/assets/apps/pipeline/sections/Mapping.jsx]: I found this approach so "jquery" approach and we are using "react" that can take care of this in a better way :thinking: 

Maybe in the render method...

```javascript
// EntityTypeViewer.jsx

// before
// result.push(<li key={field.name}><span className='name'>{field.name}</span> ... )

// possible solution
result.push(
  <li key={field.name} className={ (field in any mapping rule) ? 'entityType-mapped' : '' }>
    <span className='name'>{field.name}</span>
  // ...
)

```

File: aether-ui/aether/ui/assets/apps/pipeline/Pipeline.jsx [hunk 3]
```diff
@@ -35,6 +35,7 @@ class Pipeline extends Component {
   }
 
   componentWillReceiveProps (nextProps) {
+    console.log(this.props.selectedPipeline)
```
Comment (obdulia-losantos) [inline aether-ui/aether/ui/assets/apps/pipeline/Pipeline.jsx]: Oops!

File: aether-ui/aether/ui/assets/tests/ui-tests-environment/UiTestsEnvironment.jsx [hunk 4]
```diff
@@ -34,6 +34,7 @@ class UiTestsEnvironment extends JSDOMEnvironment {
     this.global.window.$ = $(this.global.window)
     this.global.window.jQuery = this.global.window.$
     this.global.window.Popper = popper
+    this.global.HTMLElement = this.global.window.HTMLElement
```
Comment (obdulia-losantos) [inline aether-ui/aether/ui/assets/tests/ui-tests-environment/UiTestsEnvironment.jsx]: I forgot to remove this line :man_facepalming:

File: aether-ui/aether/ui/assets/css/base/_color-codes.scss [hunk 5]
```diff
@@ -11,9 +23,33 @@ $colors-list: #6ed7c2 #6ebcd7 #cfd76e #d7b46e #d79a6e #d76e89 #d76ec2 #a5b990 #8
...
+      background-color: rgba($current-color, .2);
+    }
+
+    &::after {
+      border-left: 0.6rem solid $current-color;
+    }
   }
 }
 
+.input-mapped {
```
Comment (obdulia-losantos) [inline aether-ui/aether/ui/assets/css/base/_color-codes.scss]: Should this one be the `.input-mapped-0`? (this happens when the input schema entry is in a rule but the Entity type is missing)

Comment (lordmallam) [reply aether-ui/aether/ui/assets/css/base/_color-codes.scss]: Is that supposed to be allowed? to have a mapping with a source but without a destination?

Comment (obdulia-losantos) [reply aether-ui/aether/ui/assets/css/base/_color-codes.scss]: No, but you can have it with a wrong destination ;-)

rule:
 - source: `participant.firstName`
 - destination: `persson.name` (wrong)

Entity types:
 - Person
 - Participant
 - ...

Comment (lordmallam) [reply aether-ui/aether/ui/assets/css/base/_color-codes.scss]: If that happens, the property on entity type will not be highlighted, there should only be an error highlight on the mapping textbox (destination)

Comment (obdulia-losantos) [reply aether-ui/aether/ui/assets/css/base/_color-codes.scss]: In that case, I think you can safely remove this entry.

File: aether-ui/aether/ui/assets/css/base/_color-codes.scss [hunk 6]
```diff
@@ -49,7 +49,3 @@ $colors-list: (
     }
   }
 }
-
-.input-mapped {
-  background-color: $light-grey;
-}
```
Comment (obdulia-losantos) [inline aether-ui/aether/ui/assets/css/base/_color-codes.scss]: :+1:

File: aether-ui/aether/ui/assets/apps/components/AvroSchemaViewer.jsx [hunk 7]
```diff
@@ -3,37 +3,73 @@ import { FormattedMessage } from 'react-intl'
 import avro from 'avsc'
 
 class AvroSchemaViewer extends Component {
-  schemaToMarkup (schema) {
-    const children = []
+  getHighlightedClassName (jsonPath) {
+    const {highlight} = this.props
+    // the simplest way (equity)
```
Comment (quesebifurcan) [inline aether-ui/aether/ui/assets/apps/components/AvroSchemaViewer.jsx]: Spelling: "equity" -> "equality"?

File: aether-ui/aether/ui/assets/apps/components/AvroSchemaViewer.jsx [hunk 8]
```diff
@@ -3,37 +3,73 @@ import { FormattedMessage } from 'react-intl'
...
+    for (let i = 0; i < keys.length; i++) {
+      if (keys[i] === jsonPath) {
+        return `input-mapped-${highlight[keys[i]]}`
+      }
+    }
+
+    return ''
+  }
+
+  schemaToMarkup (schema, parent) {
```
Comment (quesebifurcan) [inline aether-ui/aether/ui/assets/apps/components/AvroSchemaViewer.jsx]: Perhaps add a comment about the type of `parent` in this context? 
Is it a string? If so, where does it come from? Can it be `null`?

File: aether-ui/aether/ui/assets/apps/mock/schema_input.mock.jsx [hunk 9]
```diff
@@ -51,7 +51,7 @@ const inputSchema = {
       }, {
         'name': 'birthYear',
         'type': 'int',
-        'doc': `Type inferred from "1972"`
+        'doc': 'Type inferred from “1972”'
```
Comment (quesebifurcan) [inline aether-ui/aether/ui/assets/apps/mock/schema_input.mock.jsx]: I think this can be removed -- it's just a leftover from when this schema was derived.

**Review by quesebifurcan — CHANGES_REQUESTED** [review]

**Review by quesebifurcan — APPROVED** [review]

**Pull request merged**